File: 00-code(源代码)/src/com/hi/dhl/algorithms/leetcode/_200/java/Solution.java
package com.hi.dhl.algorithms.leetcode._200.java;

import java.util.LinkedList;
import java.util.Queue;

/**
 * <pre>
 *     author: dhl
 *     date  : 2020/7/31
 *     desc  :
 * </pre>
 */

class Solution {

    /**
     * 广度优先遍历
     *
     * @param grid
     * @return
     */
    public int numIslands(char[][] grid) {
        if (grid == null || grid.length == 0) {
            return 0;
        }

        int row = grid.length;
        int colum = grid[0].length;
        Queue<int[]> queue = new LinkedList<>();
        int step = 0;
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < colum; j++) {
                if (grid[i][j] == '1') {
                    queue.offer(new int[]{i, j});
                    step++;
                    while (!queue.isEmpty()) {
                        int[] tmp = queue.poll();
                        int x1 = tmp[0];
                        int y1 = tmp[1];
                        if (x1 < 0 || x1 >= row || y1 < 0 || y1 >= colum || grid[x1][y1] == '0') {
                            continue;
                        }
                        grid[x1][y1] = '0';
                        queue.offer(new int[]{x1 + 1, y1});
                        queue.offer(new int[]{x1 - 1, y1});
                        queue.offer(new int[]{x1, y1 + 1});
                        queue.offer(new int[]{x1, y1 - 1});
                    }
                }

            }
        }
        return step;
    }


    /**
     * 深度优先遍历
     *
     * @param grid
     * @return
     */
    public int numIslands2(char[][] grid) {
        if (grid == null || grid.length == 0) {
            return 0;
        }
        int row = grid.length;
        int colum = grid[0].length;
        int step = 0;
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < colum; j++) {
                if (grid[i][j] == '1') {
                    step++;
                    dfs(i, j, row, colum, grid);
                }

            }
        }
        return step;
    }

    void dfs(int x1, int y1, int row, int colum, char[][] grid) {
        if (x1 < 0 || x1 >= row || y1 < 0 || y1 >= colum || grid[x1][y1] == '0') {
            return;
        }
        grid[x1][y1] = '0';
        dfs(x1 + 1, y1, row, colum, grid);
        dfs(x1 - 1, y1, row, colum, grid);
        dfs(x1, y1 + 1, row, colum, grid);
        dfs(x1, y1 - 1, row, colum, grid);
    }
}

File: 00-code(源代码)/src/com/hi/dhl/algorithms/leetcode/_200/kotlin/Solution.kt
package com.hidhl.leetcode._200.kotlin

import java.util.*

/**
 * <pre>
 *     author: dhl
 *     date  : 2020/7/31
 *     desc  :
 * </pre>
 */

class Solution {

    /**
     * 广度优先遍历
     */
    fun numIslands(grid: Array<CharArray>): Int {
        if (grid.size == 0) {
            return 0;
        }

        var step = 0;
        val queue = LinkedList<IntArray>()
        val row = grid.size
        val colum = grid[0].size
        for (i in 0 until row) {
            for (j in 0 until colum) {
                if (grid[i][j] == '1') {
                    step = step + 1
                    queue.offer(intArrayOf(i, j))
                    while (!queue.isEmpty()) {
                        val (x1, y1) = queue.poll()
                        if (x1 < 0 || x1 >= row || y1 < 0 || y1 >= colum || grid[x1][y1] == '0') {
                            continue;
                        }

                        grid[x1][y1] = '0'
                        queue.offer(intArrayOf(x1 + 1, y1))
                        queue.offer(intArrayOf(x1 - 1, y1))
                        queue.offer(intArrayOf(x1, y1 + 1))
                        queue.offer(intArrayOf(x1, y1 - 1))
                    }
                }
            }
        }
        return step
    }


    /**
     * 深度优先遍历
     */
    fun numIslands2(grid: Array<CharArray>): Int {
        if (grid.size == 0) {
            return 0;
        }

        var step = 0;
        val row = grid.size
        val colum = grid[0].size
        for (i in 0 until row) {
            for (j in 0 until colum) {
                if (grid[i][j] == '1') {
                    step = step + 1
                    dfs(i, j, row, colum, grid)
                }
            }
        }
        return step
    }

    fun dfs(x1: Int, y1: Int, row: Int, colum: Int, grid: Array<CharArray>) {
        if (x1 < 0 || x1 >= row || y1 < 0 || y1 >= colum || grid[x1][y1] == '0') {
            return;
        }

        grid[x1][y1] = '0'
        dfs(x1 + 1, y1, row, colum, grid)
        dfs(x1 - 1, y1, row, colum, grid)
        dfs(x1, y1 + 1, row, colum, grid)
        dfs(x1, y1 - 1, row, colum, grid)
    }
}